### PR TITLE
sd-boot: clear NV variables only if present

### PR DESCRIPTION
**Pull request by evelikov**
We already do so for some NV variables - see config_load_defaults(). Just use the same approach across the board - check before nuking.

### CODE REVIEW

File: src/boot/efi/boot.c [hunk 1]
```diff
@@ -1057,18 +1057,30 @@ static bool menu_run(
                 efivar_set(MAKE_GUID_PTR(LOADER), u"LoaderEntryDefault", config->entry_default_efivar, EFI_VARIABLE_NON_VOLATILE);
 
         if (console_mode_efivar_saved != config->console_mode_efivar) {
-                if (config->console_mode_efivar == CONSOLE_MODE_KEEP)
-                        efivar_set(MAKE_GUID_PTR(LOADER), u"LoaderConfigConsoleMode", NULL, EFI_VARIABLE_NON_VOLATILE);
-                else
+                if (config->console_mode_efivar == CONSOLE_MODE_KEEP) {
+                        _cleanup_free_ char *buf = NULL;
```
Comment (bluca) [inline src/boot/efi/boot.c]: given you are not using the return values, just pass NULL for both, you'll still get the success/failure

**Review by bluca — APPROVED** [review]

File: src/boot/efi/boot.c (hunk 1 above)
Comment (evelikov) [reply src/boot/efi/boot.c]: Neat, PR updated.

**Comment by poettering** [comment]
what's the rationale here? i mean, if we delete a variable that doesn't exist it should return an error, which is fine.

If there's a reason to do this then I'd prefer if we'd add a proper function instead for deleting vars, i.e. efivar_unset() or so, that hides this mess inside.

**Comment by poettering** [comment]
Yeah, efivar_unset() is definitely the way to go I think. In fact the weird way to delete efu vars by calling efivar_set() with a NULL pointer should be removed...

**Comment by medhefgo** [comment]
Exit code is `EFI_NOT_FOUND` if "The variable trying to be updated or deleted was not found.". A sane person would deduce no write to be incurred. But firmware authors *mumblemumble*…

File: src/boot/efi/util.c [hunk 2]
```diff
@@ -85,6 +85,19 @@ EFI_STATUS efivar_set_uint64_le(const EFI_GUID *vendor, const char16_t *name, ui
         return efivar_set_raw(vendor, name, buf, sizeof(buf), flags);
 }
 
+EFI_STATUS efivar_unset(const EFI_GUID *vendor, const char16_t *name, uint32_t flags) {
+        EFI_STATUS err;
+
+        assert(vendor);
+        assert(name);
+
+        err = efivar_get_raw(vendor, name, NULL, NULL);
```
Comment (medhefgo) [inline src/boot/efi/util.c:96]: A comment as to why would be nice.

**Review by bluca — APPROVED** [review]

File: src/boot/efi/util.c [hunk 3]
```diff
@@ -85,6 +85,19 @@ EFI_STATUS efivar_set_uint64_le(const EFI_GUID *vendor, const char16_t *name, ui
...
+        if (err == EFI_SUCCESS)
+                err = efivar_set_raw(vendor, name, NULL, 0, flags);
```
Comment (medhefgo) [inline src/boot/efi/util.c]: (We prefer early returns.)
```suggestion
                return efivar_set_raw(vendor, name, NULL, 0, flags);
```

File: src/boot/efi/util.c (hunk 2 above)
Comment (evelikov) [reply src/boot/efi/util.c:96]: Done, thanks.

**Review by medhefgo — APPROVED** [review]

File: src/boot/efi/util.c [hunk 4]
```diff
@@ -85,6 +85,21 @@ EFI_STATUS efivar_set_uint64_le(const EFI_GUID *vendor, const char16_t *name, ui
         return efivar_set_raw(vendor, name, buf, sizeof(buf), flags);
 }
 
+EFI_STATUS efivar_unset(const EFI_GUID *vendor, const char16_t *name, uint32_t flags) {
+        EFI_STATUS err;
+
+        assert(vendor);
+        assert(name);
+
+        /* We could be wiping a non-volatile variable here and the spec makes no guarantees that won't incurr
```
Comment (bluca) [inline src/boot/efi/util.c]: incurr -> incur
and extra write -> in an extra write
as needed -> if needed

**Pull request merged**